internal/process/snapshot.go

```diff
@@ -0,0 +1,124 @@
+
+	info = map[types.ManagedObjectReference]*infoSnapshot{}
+	if ex != nil {
+		for _, exFile := range ex.File {
+			if exFile.Type == "snapshotData" {
+				findSnapshotAndUpdateInfoData(ex.Snapshot, exFile, info)
+			} else if exFile.Type == "snapshotMemory" {
+				findSnapshotAndUpdateInfoMemory(ex.Snapshot, exFile, info)
+			} else if exFile.Type == "suspendMemory" {
+				suspendMemory += exFile.Size
+				suspendMemoryUnique += exFile.UniqueSize
+			}
+		}
+	}
+	return info, suspendMemory, suspendMemoryUnique
+}
+
+// It finds for the given datakey the snapshotRef and saves the data
+func findSnapshotAndUpdateInfoData(snapshostFiles []types.VirtualMachineFileLayoutExSnapshotLayout, exFile types.VirtualMachineFileLayoutExFileInfo, info map[types.ManagedObjectReference]*infoSnapshot) {
+	for _, exs := range snapshostFiles {
+		if exs.DataKey == exFile.Key {
+			i, ok := info[exs.Key]
+			if !ok {
+				i = &infoSnapshot{}
+				info[exs.Key] = i
+			}
+			i.totalDisk += exFile.Size
+			i.totalUniqueDisk += exFile.UniqueSize
+			i.datastorePathDisk = i.datastorePathDisk + exFile.Name + "|"
+		}
+	}
+}
+
+// It finds for the given memorykey the snapshotRef and saves the data
+func findSnapshotAndUpdateInfoMemory(snapshostFiles []types.VirtualMachineFileLayoutExSnapshotLayout, exFile types.VirtualMachineFileLayoutExFileInfo, info map[types.ManagedObjectReference]*infoSnapshot) {
+	for _, exs := range snapshostFiles {
+		if exs.MemoryKey == exFile.Key && exs.MemoryKey != -1 {
+			i, ok := info[exs.Key]
+			if !ok {
+				i = &infoSnapshot{}
+				info[exs.Key] = i
+			}
+			i.totalMemoryInDisk += exFile.Size
+			i.totalUniqueMemoryInDisk += exFile.UniqueSize
+			i.datastorePathMemory = i.datastorePathMemory + exFile.Name + "|"
+		}
+	}
+}
+
+// It adds a new sample for each snapshot following the whole tree in a recursive way
+func traverseSnapshotList(e *integration.Entity, config *load.Config, tree types.VirtualMachineSnapshotTree, treeInfo string, info map[types.ManagedObjectReference]*infoSnapshot) {
+
+	ms := e.NewMetricSet("VSphere" + entityTypeSnapshotVm + "Sample")
+	treeInfo = treeInfo + ":" + tree.Name
+	createMetricsCurrentSnapshot(treeInfo, tree, config, ms, info)
+
+	//A recursive function is needed since the actual size of the tree in unknown
+	for _, s := range tree.ChildSnapshotList {
+		traverseSnapshotList(e, config, s, treeInfo, info)
+	}
+
+}
+
+func createMetricsCurrentSnapshot(treeInfo string, tree types.VirtualMachineSnapshotTree, config *load.Config, ms *metric.Set, info map[types.ManagedObjectReference]*infoSnapshot) {
+	checkError(config, ms.SetMetric("snapshotTreeInfo", treeInfo, metric.ATTRIBUTE))
+	checkError(config, ms.SetMetric("name", tree.Name, metric.ATTRIBUTE))
+	checkError(config, ms.SetMetric("creationTime", tree.CreateTime.String(), metric.ATTRIBUTE))
+	checkError(config, ms.SetMetric("powerState", string(tree.State), metric.ATTRIBUTE))
```

Comment:
should't we add the powerState(in an aggregated way) in the vm event in order to alert on vm that are running from snapshots?

Reply:
the power state attributes indicates the status of the machine when the snapshot was taken, not the power status of the snapshot